### PR TITLE
`Feature`: Modeling exercise feedback

### PR DESCRIPTION
**Pull request by matthiaslehnertum**
### Motivation and Context
As Athena so far does not support providing feedback on modeling exercises, this PR adds another module to allow doing so.

### Description
This pull request adds support for providing feedback on modeling exercises to Athena. While the module should support all diagram types supported in Apollon, currently only the BPMN diagram type has been tested and optimised. Additionally, the playground is extended to allow previewing of modeling-submissions within the playground.

The following table lists the recognition performance of different modeling errors in BPMN diagrams:

| **Testcase** | **Error correctly identified?** |
| :------------ | :---- |
| 1.1.1.1 Missing start event | no |
| 1.1.1.2 Missing end event | yes |
| 1.2.1 Non-control-flow as control-flow | no |
| 1.2.2.1 Using an event as indicator | no |
| 1.2.2.2 Wrong use of start event | no |
| 1.2.2.3 Wrong use of end event | no |
| 1.2.3.1.1 Unlabelled activity | yes |
| 1.2.3.1.2 Unlabelled gateway | yes |
| 1.2.3.1.3 Unlabelled event | no |
| 1.2.4.1.1 No gateway logic description | no |
| 1.3.1 Missing incoming connection | no |
| 1.3.2 Missing outgoing connection | no |
| 1.4.1.1 Sequence flow crossing pool boundaries | no |
| 1.4.1.2 Unlabelled pool | yes |
| 1.4.2.1 Unlabelled lane | yes |
| 2.1.1.1.1 Activity missing entirely | partially |
| 2.1.1.1.2 Activity has been merged | partially |
| 2.1.1.2 Missing event | partially |
| 2.1.1.2 Missing gateway | partially |
| 2.1.2.1 Missing sequence flow | partially |
| 2.1.2.2 Missing message flow | no |
| 2.1.2.3 Missing association | no |
| 2.2.1 Use of wrong activity type | no |
| 2.2.2 Use of wrong event type | no |
| 2.2.3 Use of wrong gateway type | no |
| 2.3.1.1 Wrongly named activity | partially |
| 2.3.1.2 Wrongly named event | yes |
| 2.3.1.3 Wrongly named gateway | yes |
| 2.3.2 Wrongly named connecting objects | no |
| 2.3.3 Wrongly named pools, lanes and artifacts | yes |
| 2.4 Excessive diagram object | yes |
| 2.5 Wrong sequence of diagram objects | yes |
| 2.6.1 Wrong use of the sequential multi instance marker | no |
| 2.6.2 Wrong use of the parallel multi instance marker | no |
| 2.6.3 Wrong use of the loop marker | no |
| 3.1.1 Use of uncontrolled flows | yes |
| 3.2.1.1 Missing description on timed event | no |
| 3.3.1.1 Wrong naming scheme | yes |

### Steps for Testing
- Step 1: Start the assessment module manager and the module "module_modelling_llm".
- Step 2: Start the playground
- Step 3: Open localhost:3000 in your browser
- Step 4: Select the module "module_modelling_llm" in the playground
- Step 5: Select the exercise "6 modelling: Process Modelling with BPMN: Obtaining a loan"
- Step 6: Send the submissions to Athena and request submission selection for them
- Step 7: Select the exercise "6 modeling: Process Modelling with BPMN: Obtaining a loan" again in the section "Request Feedback Suggestions from Athena" and then select for example "3.3.1.1 Wrong name scheme"
- Step 8: When clicking the button "Request feedback selection", Athena should show a list of feedback items corresponding to the selected submission. For the submission "3.3.1.1 Wrong name scheme", valid feedback would be that the activities in the diagram are not named according to the "Verb Object" format referenced in the grading instructions.

Screenshot 2024-01-21 at 17 06 56

### Screenshots
Modeling submission preview in the playground:

Screenshot 2024-01-21 at 16 56 56

### CODE REVIEW

File: playground/src/model/feedback.ts [hunk 1]
```diff
@@ -32,7 +32,11 @@ export type ProgrammingFeedback = FeedbackBase & {
   line_end?: number;
 };
 
-export type Feedback = TextFeedback | ProgrammingFeedback;
+export type ModellingFeedback = FeedbackBase & {
+  type: "modelling";
```
Comment (maximiliansoelch) [inline playground/src/model/feedback.ts:36]: Shouldn't there be some kind of reference to the model element the feedback is assigned to?

File: module_modelling_llm/README.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
+# Module Text LLM
```
Comment (maximiliansoelch) [inline module_modelling_llm/README.md]: This should be adapted for the modeling module

File: module_modelling_llm/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,29 @@
+# syntax=docker/dockerfile:1
+
+# This is the Dockerfile for the assessment module manager.
```
Comment (maximiliansoelch) [inline module_modelling_llm/Dockerfile]: Copy&Paste error

File: assessment_module_manager/modules.docker.ini [hunk 4]
```diff
@@ -32,3 +32,8 @@ supports_evaluation = false
 url = http://module-programming-code-embedding:5007
 type = programming
 supports_evaluation = false
+
+[module_modelling_llm]
+url = http://module-modelling-llm:5008
+type = modelling
+supports_evaluation = false
```
Comment (maximiliansoelch) [inline assessment_module_manager/modules.docker.ini:39]: Shouldn't this be true? 🤔
```suggestion
supports_evaluation = true
```

File: assessment_module_manager/modules.ini [hunk 5]
```diff
@@ -33,3 +33,7 @@ url = http://localhost:5007
 type = programming
 supports_evaluation = false
 
+[module_modelling_llm]
+url = http://localhost:5008
+type = modelling
+supports_evaluation = false
```
Comment (maximiliansoelch) [inline assessment_module_manager/modules.ini:39]: Same as above:
```suggestion
supports_evaluation = true
```

File: module_modelling_llm/module_modelling_llm/helpers/models/replicate.py
Comment (maximiliansoelch) [inline module_modelling_llm/module_modelling_llm/helpers/models/replicate.py:1]: Do we already need this file?

**Review by maximiliansoelch — CHANGES_REQUESTED** [review]

**Review by marlon-luca-bu — APPROVED** [review]
Code looks good to me 👍

**Review by maximiliansoelch — APPROVED** [review]
Changes look good, thanks for the contribution 👍

**Pull request merged**